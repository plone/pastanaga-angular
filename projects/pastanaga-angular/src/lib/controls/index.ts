export * from './control.model';
export * from './form-field.model';
export * from './form-field.utils';

export * from './form-field';
export * from './textfield';
export * from './toggles';
export * from './slider';
