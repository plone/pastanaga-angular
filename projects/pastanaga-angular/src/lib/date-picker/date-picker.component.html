<div [paPopup]="popup" popupOnRight="true" #popupRef="paPopupRef">
    <pa-input
        #input
        size="small"
        icon="calendar"
        iconOnRight
        [readonly]="readonly"
        [formControl]="inputControl"
        (click)="handleInputClick($event)"
        >{{ label || 'Pick a date' }}</pa-input
    >
</div>
<pa-popup [ngSwitch]="mode" #popup (onOpen)="openWeeks()" (onClose)="handlePopupClose()" class="calendar-popup">
    <ng-container *ngSwitchCase="'weeks'">
        <div class="calendar-header">
            <button (click)="openYears($event)">
                {{ trackedDate | date: 'yyyy' }}
                <pa-icon name="chevron-down"></pa-icon>
            </button>
            <button (click)="openMonths($event)">
                {{ trackedDate | date: 'LLLL' }}
                <pa-icon name="chevron-down"></pa-icon>
            </button>
            <div class="spacer"></div>
            <pa-button
                (click)="previousMonth()"
                kind="secondary"
                aspect="basic"
                size="small"
                icon="chevron-left"
            ></pa-button>
            <pa-button
                (click)="nextMonth()"
                kind="secondary"
                aspect="basic"
                size="small"
                icon="chevron-right"
            ></pa-button>
        </div>
        <div class="calendar-weekdays">
            <div *ngFor="let day of days">{{ day }}</div>
        </div>
        <div *ngFor="let week of weeks" class="calendar-row">
            <pa-button
                *ngFor="let day of week"
                (click)="selectDay(day)"
                size="small"
                aspect="basic"
                [class.other-month]="day.otherMonth"
                [class.selected]="day.selected"
                >{{ day.number }}</pa-button
            >
        </div>
    </ng-container>
    <ng-container *ngSwitchCase="'years'">
        <div class="calendar-list-header overline">
            <div class="label">Years</div>
            <pa-button
                (click)="openWeeks($event)"
                kind="secondary"
                aspect="basic"
                size="small"
                icon="chevron-left"
            ></pa-button>
        </div>
        <div class="calendar-list">
            <div
                *ngFor="let year of years"
                (click)="selectYear($event, year.number)"
                class="calendar-list-item"
                [class.selected]="year.selected"
            >
                {{ year.number }}
            </div>
        </div>
    </ng-container>
    <ng-container *ngSwitchCase="'months'">
        <div class="calendar-list-header overline">
            <div class="label">Months</div>
            <pa-button
                (click)="openWeeks($event)"
                kind="secondary"
                aspect="basic"
                size="small"
                icon="chevron-left"
            ></pa-button>
        </div>
        <div class="calendar-list">
            <div
                *ngFor="let month of months; index as i"
                (click)="selectMonth($event, i)"
                class="calendar-list-item"
                [class.selected]="month.selected"
            >
                {{ month.name }}
            </div>
        </div>
    </ng-container>
</pa-popup>
