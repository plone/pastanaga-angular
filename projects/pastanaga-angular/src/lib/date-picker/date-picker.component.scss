@import '../../styles/variables';

:host {
    --day-width: 32px;
    --header-color: #{$color-neutral-strong};
    --text-color: #{$color-dark-stronger};
    --text-color-other-month: #{$color-neutral-light};
    --weekday-border-color: #{$color-neutral-light};
    --selected-background-color: #{$color-primary-lightest};
    --list-item-background-color-hover: #{$color-neutral-lighter};

    .calendar-popup {
        font-size: 14px;
        line-height: 24px;
        color: var(--text-color);

        ::ng-deep {
            .pa-popup {
                width: 320px;
            }
        }
    }

    .calendar-header {
        display: flex;
        align-items: center;

        button {
            border: none;
            background: none;
            outline: 0;

            &:focus {
                box-shadow: $shadow-focus !important;
            }
        }

        .spacer {
            flex-grow: 1;
        }
    }

    .calendar-weekdays,
    .calendar-row {
        display: flex;
        color: var(--header-color);
        padding: 6px 0;
        justify-content: space-around;
    }

    .calendar-weekdays {
        border-bottom: 1px solid var(--weekday-border-color);

        div {
            width: var(--day-width);
            text-align: center;
        }
    }

    .calendar-row ::ng-deep pa-button {
        button {
            width: var(--day-width);
        }

        &.other-month,
        &.other-month:hover {
            button {
                color: var(--text-color-other-month);
            }
        }
    }

    .calendar-list-header {
        display: flex;
        align-items: center;
        color: var(--header-color);
        font-size: 12px;
        font-weight: 700;
        line-height: 16px;
        text-transform: uppercase;
        padding: 20px 12px 12px;

        .label {
            flex-grow: 1;
        }
    }

    .calendar-list-item {
        padding: 12px;
        font-size: 16px;
        cursor: pointer;

        &.selected {
            background-color: var(--selected-background-color);
        }

        &:hover,
        &.selected:hover {
            background-color: var(--list-item-background-color-hover);
        }
    }

    .calendar-row ::ng-deep pa-button.selected {
        background-color: var(--selected-background-color);
        border-radius: 100px;
    }
}
