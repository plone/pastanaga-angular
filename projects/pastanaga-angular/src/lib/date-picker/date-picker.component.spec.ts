import { ReactiveFormsModule } from '@angular/forms';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { addMinutes, addMonths, parse, setMonth, setYear, subMonths } from 'date-fns';
import { MockModule } from 'ng-mocks';
import { DatePickerComponent, Day } from './date-picker.component';
import { PaButtonModule } from '../button';
import { PaIconModule } from '../icon';
import { PaPopupModule } from '../popup';
import { PaTextFieldModule } from '../controls';
import { markForCheck, TRANSITION_DURATION } from '../common';
import { formatDate } from '@angular/common';
import { PaTranslateModule } from '../translate';

jest.mock('../common', () => ({
    markForCheck: jest.fn(),
    TRANSITION_DURATION: {
        moderate: 500,
    },
}));

describe('DatePickerComponent', () => {
    let component: DatePickerComponent;
    let componentAny: any;

    const createComponent = createComponentFactory({
        component: DatePickerComponent,
        imports: [
            MockModule(PaButtonModule),
            MockModule(PaIconModule),
            MockModule(PaPopupModule),
            MockModule(PaTextFieldModule),
            MockModule(PaTranslateModule),
            ReactiveFormsModule,
        ],
    });

    let spectator: Spectator<DatePickerComponent>;

    beforeEach(() => {
        spectator = createComponent();
        componentAny = component = spectator.component;
        jest.useFakeTimers();
    });

    afterEach(() => {
        jest.useRealTimers();
    });

    it('should create', () => {
        expect(component).toBeTruthy();
    });

    describe('value control changes', () => {
        beforeEach(() => {
            component.inputControl.setValue = jest.fn();
        });

        it('should do nothing for null value', () => {
            // === Execute ===
            component.control.setValue(null);

            // === Verify ===
            expect(component.inputControl.setValue).not.toHaveBeenCalled();
        });

        it('should update input control with value', () => {
            // === Setup ===
            const text = 'March 22, 1980';
            const date = formatDate(text, 'longDate', componentAny.locale);

            // === Execute ===
            component.control.setValue(text);

            // === Verify ===
            expect(component.inputControl.setValue).toHaveBeenCalledWith(date);
        });
    });

    describe('input control changes', () => {
        beforeEach(() => {
            componentAny.generateWeeks = jest.fn();
            componentAny.setDate = jest.fn();
        });

        it('should debounce', () => {
            // === Execute ===
            component.inputControl.setValue(null);

            // === Verify ===
            expect(componentAny.generateWeeks).not.toHaveBeenCalled();

            // === Execute ===
            jest.advanceTimersByTime(TRANSITION_DURATION.moderate);

            // === Verify ===
            expect(componentAny.setDate).toHaveBeenCalled();
            expect(componentAny.generateWeeks).toHaveBeenCalled();
            expect(markForCheck).toHaveBeenCalledWith(componentAny.cdr);
        });

        it('should clear date when there is no value', () => {
            // === Setup ===
            componentAny._selectedDate = new Date();
            component.trackedDate = new Date('2020-01-03');
            const date = new Date();
            jest.setSystemTime(date);

            // === Execute ===
            component.inputControl.setValue(null);
            jest.advanceTimersByTime(TRANSITION_DURATION.moderate);

            // === Verify ===
            expect(componentAny.setDate).toHaveBeenCalledWith(undefined);
            expect(componentAny.generateWeeks).toHaveBeenCalled();
            expect(markForCheck).toHaveBeenCalledWith(componentAny.cdr);
            expect(component.trackedDate.getTime()).toEqual(date.getTime() + TRANSITION_DURATION.moderate);
        });

        it('should clear date for invalid value', () => {
            // === Execute ===
            component.inputControl.setValue('foobar');
            jest.advanceTimersByTime(TRANSITION_DURATION.moderate);

            // === Verify ===
            expect(componentAny.setDate).toHaveBeenCalledWith(undefined);
            expect(componentAny.generateWeeks).toHaveBeenCalled();
            expect(markForCheck).toHaveBeenCalledWith(componentAny.cdr);
        });

        it('should update with date', () => {
            // === Setup ===
            const value = 'March 22, 2022';
            const expected = parse(value, "MMMM d',' yyyy", new Date());

            // === Execute ===
            component.inputControl.setValue(value);
            jest.advanceTimersByTime(TRANSITION_DURATION.moderate);

            // === Verify ===
            expect(componentAny.setDate).toHaveBeenCalledWith(expected);
            expect(componentAny.generateWeeks).toHaveBeenCalled();
            expect(markForCheck).toHaveBeenCalledWith(componentAny.cdr);
        });
    });

    describe('disabled field ', () => {
        it('should set disabled state using string', () => {
            // === Setup ===
            component.setDisabledState = jest.fn();

            // === Execute ===
            component.disabled = 'false';

            // === Verify ===
            expect(component.setDisabledState).toHaveBeenCalledWith(false);
        });

        it('should set disabled state using boolean', () => {
            // === Setup ===
            component.setDisabledState = jest.fn();

            // === Execute ===
            component.disabled = true;

            // === Verify ===
            expect(component.setDisabledState).toHaveBeenCalledWith(true);
        });

        it('should get value from control', () => {
            // === Setup ===
            const getter = jest.spyOn(component.inputControl, 'disabled', 'get').mockReturnValue(true);

            // === Verify ===
            expect(component.disabled).toBe(true);
            expect(getter).toHaveBeenCalled();
        });
    });

    it('should generate weeks', () => {
        // === Setup ===
        component.trackedDate = new Date('2022-7-10');

        // === Execute ===
        componentAny.generateWeeks();

        // === Verify ===
        expect(component.weeks).toMatchSnapshot();
    });

    it('should generate months', () => {
        // === Setup ===
        component.trackedDate = new Date('2022-7-10');

        // === Execute ===
        componentAny.generateMonths();

        // === Verify ===
        expect(component.months).toMatchSnapshot();
    });

    describe('generating years', () => {
        beforeEach(() => {
            jest.setSystemTime(new Date('2022-03-22'));
        });

        it('should max at current year when tracked is earlier', () => {
            // === Setup ===
            component.trackedDate = new Date('2015-4-9');

            // === Execute ===
            componentAny.generateYears();

            // === Verify ===
            expect(component.years).toMatchSnapshot();
        });

        it('should max at tracked year when tracked is later', () => {
            // === Setup ===
            component.trackedDate = new Date('2040-4-9');

            // === Execute ===
            componentAny.generateYears();

            // === Verify ===
            expect(component.years).toMatchSnapshot();
        });

        it('should min at 2000 year when tracked is later', () => {
            // === Setup ===
            component.trackedDate = new Date('2002-4-9');

            // === Execute ===
            componentAny.generateYears();

            // === Verify ===
            expect(component.years).toMatchSnapshot();
        });

        it('should min at tracked year when tracked is earlier', () => {
            // === Setup ===
            component.trackedDate = new Date('1998-4-9');

            // === Execute ===
            componentAny.generateYears();

            // === Verify ===
            expect(component.years).toMatchSnapshot();
        });
    });

    describe('input click handler', () => {
        let event: MouseEvent;

        beforeEach(() => {
            event = {
                stopPropagation: jest.fn(),
                preventDefault: jest.fn(),
            } as any;
        });

        it('should trap events when popup is displayed', () => {
            // === Setup ===
            component.popup = { isDisplayed: true } as any;
            componentAny._touched = false;
            jest.spyOn(component, 'disabled', 'get').mockReturnValue(false);
            component.readonly = false;
            const touched = jest.fn();
            component.registerOnTouched(touched);

            // === Execute ===
            component.handleInputClick(event);

            // === Verify ===
            expect(event.stopPropagation).toHaveBeenCalled();
            expect(event.preventDefault).toHaveBeenCalled();
            expect(touched).toHaveBeenCalled();
        });

        it('should trap events when input is disabled', () => {
            // === Setup ===
            component.popup = { isDisplayed: false } as any;
            componentAny._touched = false;
            jest.spyOn(component, 'disabled', 'get').mockReturnValue(true);
            component.readonly = false;
            const touched = jest.fn();
            component.registerOnTouched(touched);

            // === Execute ===
            component.handleInputClick(event);

            // === Verify ===
            expect(event.stopPropagation).toHaveBeenCalled();
            expect(event.preventDefault).toHaveBeenCalled();
            expect(touched).not.toHaveBeenCalled();
        });

        it('should trap events when input is readonly', () => {
            // === Setup ===
            component.popup = { isDisplayed: false } as any;
            componentAny._touched = false;
            jest.spyOn(component, 'disabled', 'get').mockReturnValue(false);
            component.readonly = true;
            const touched = jest.fn();
            component.registerOnTouched(touched);

            // === Execute ===
            component.handleInputClick(event);

            // === Verify ===
            expect(event.stopPropagation).toHaveBeenCalled();
            expect(event.preventDefault).toHaveBeenCalled();
            expect(touched).not.toHaveBeenCalled();
        });

        it('should not trap events when popup is not displayed', () => {
            // === Setup ===
            component.popup = { isDisplayed: false } as any;
            const touched = jest.fn();
            component.registerOnTouched(touched);

            // === Execute ===
            component.handleInputClick(event);

            // === Verify ===
            expect(event.stopPropagation).not.toHaveBeenCalled();
            expect(event.preventDefault).not.toHaveBeenCalled();
            expect(touched).toHaveBeenCalled();
        });
    });

    describe('popup close handler', () => {
        it('should set tracked date to selected date', () => {
            // === Setup ===
            componentAny._selectedDate = new Date();

            // === Execute ===
            component.handlePopupClose();

            // === Verify ===
            expect(component.trackedDate).toBe(componentAny._selectedDate);
        });

        it('should set tracked date to a new date', () => {
            // === Setup ===
            const date = new Date();
            jest.setSystemTime(date);

            // === Execute ===
            component.handlePopupClose();

            // === Verify ===
            expect(component.trackedDate).toEqual(date);
        });
    });

    it('should open weeks', () => {
        // === Setup ===
        const event: Event = { stopPropagation: jest.fn() } as any;
        componentAny.generateWeeks = jest.fn();

        // === Execute ===
        component.openWeeks(event);

        // === Verify ===
        expect(event.stopPropagation).toHaveBeenCalled();
        expect(componentAny.generateWeeks).toHaveBeenCalled();
        expect(component.mode).toBe('weeks');
    });

    it('should open years', () => {
        // === Setup ===
        const event: Event = { stopPropagation: jest.fn() } as any;
        componentAny.generateYears = jest.fn();

        // === Execute ===
        component.openYears(event);

        // === Verify ===
        expect(event.stopPropagation).toHaveBeenCalled();
        expect(componentAny.generateYears).toHaveBeenCalled();
        expect(component.mode).toBe('years');
    });

    it('should open months', () => {
        // === Setup ===
        const event: Event = { stopPropagation: jest.fn() } as any;
        componentAny.generateMonths = jest.fn();

        // === Execute ===
        component.openMonths(event);

        // === Verify ===
        expect(event.stopPropagation).toHaveBeenCalled();
        expect(componentAny.generateMonths).toHaveBeenCalled();
        expect(component.mode).toBe('months');
    });

    it('should update to previous month', () => {
        // === Setup ===
        const date = new Date();
        component.trackedDate = date;
        componentAny.generateWeeks = jest.fn();

        // === Execute ===
        component.previousMonth();

        // === Verify ===
        expect(component.trackedDate).toEqual(subMonths(date, 1));
        expect(componentAny.generateWeeks).toHaveBeenCalled();
    });

    it('should update to next month', () => {
        // === Setup ===
        const date = new Date();
        component.trackedDate = date;
        componentAny.generateWeeks = jest.fn();

        // === Execute ===
        component.nextMonth();

        // === Verify ===
        expect(component.trackedDate).toEqual(addMonths(date, 1));
        expect(componentAny.generateWeeks).toHaveBeenCalled();
    });

    it('should select year', () => {
        // === Setup ===
        component.trackedDate = new Date('2022-05-02');
        component.openWeeks = jest.fn();
        const event = {} as Event;

        // === Execute ===
        component.selectYear(event, 1980);

        // === Verify ===
        expect(component.trackedDate).toEqual(setYear(component.trackedDate, 1980));
        expect(component.openWeeks).toHaveBeenCalledWith(event);
    });

    it('should select month', () => {
        // === Setup ===
        component.trackedDate = new Date('2022-05-02');
        component.openWeeks = jest.fn();
        const event = {} as Event;

        // === Execute ===
        component.selectMonth(event, 11);

        // === Verify ===
        expect(component.trackedDate).toEqual(setMonth(component.trackedDate, 11));
        expect(component.openWeeks).toHaveBeenCalledWith(event);
    });

    it('should select day', () => {
        // === Setup ===
        component.popupDirective!.toggle = jest.fn();
        const setDate = (componentAny.setDate = jest.fn());
        const day = {
            date: new Date(),
        } as Day;
        component.input = {
            htmlInputRef: {
                nativeElement: {
                    focus: jest.fn(),
                },
            },
        } as any;

        // === Execute ===
        component.selectDay(day);

        // === Verify ===
        expect(setDate).toHaveBeenCalledWith(day.date);
        expect(component.input?.htmlInputRef?.nativeElement.focus).toHaveBeenCalled();
    });

    describe('setting date', () => {
        beforeEach(() => {
            component.onChange = jest.fn();
        });

        it('should use date value', () => {
            // === Setup ===
            const date = new Date();
            const expected = addMinutes(date, date.getTimezoneOffset() * -1);

            // === Execute ===
            componentAny.setDate(date);

            // === Verify ===
            expect(componentAny._selectedDate).toEqual(expected);
            expect(component.trackedDate).toEqual(expected);
            expect(component.onChange).toHaveBeenCalledWith(expected);
        });

        it('should handle undefined', () => {
            // === Setup ===
            const expected = new Date();
            jest.setSystemTime(expected);

            // === Execute ===
            componentAny.setDate(undefined);

            // === Verify ===
            expect(componentAny._selectedDate).toEqual(undefined);
            expect(component.trackedDate).toEqual(expected);
            expect(component.onChange).toHaveBeenCalledWith(undefined);
        });
    });

    it('should set disabled state', () => {
        // === Execute ===
        component.setDisabledState(true);

        // === Verify ===
        expect(component.disabled).toBe(true);
    });

    describe('setting disabled state', () => {
        it('should do nothing if already enabled', () => {
            // === Setup ===
            component.setDisabledState(false);
            component.inputControl.enable = jest.fn();

            // === Execute ===
            component.setDisabledState(false);

            // === Verify ===
            expect(component.inputControl.enable).not.toHaveBeenCalled();
        });

        it('should do nothing if already disabled', () => {
            // === Setup ===
            component.setDisabledState(true);
            component.inputControl.disable = jest.fn();

            // === Execute ===
            component.setDisabledState(true);

            // === Verify ===
            expect(component.inputControl.disable).not.toHaveBeenCalled();
        });

        it('should become disabled', () => {
            // === Setup ===
            component.setDisabledState(false);
            component.inputControl.disable = jest.fn();

            // === Execute ===
            component.setDisabledState(true);

            // === Verify ===
            expect(component.inputControl.disable).toHaveBeenCalled();
        });

        it('should become enabled', () => {
            // === Setup ===
            component.setDisabledState(true);
            component.inputControl.enable = jest.fn();

            // === Execute ===
            component.setDisabledState(false);

            // === Verify ===
            expect(component.inputControl.enable).toHaveBeenCalled();
        });
    });
});
