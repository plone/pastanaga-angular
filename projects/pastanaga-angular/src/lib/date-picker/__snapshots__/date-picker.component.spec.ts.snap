// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePickerComponent generating years should max at current year when tracked is earlier 1`] = `
Array [
  Object {
    "number": 2022,
    "selected": false,
  },
  Object {
    "number": 2021,
    "selected": false,
  },
  Object {
    "number": 2020,
    "selected": false,
  },
  Object {
    "number": 2019,
    "selected": false,
  },
  Object {
    "number": 2018,
    "selected": false,
  },
  Object {
    "number": 2017,
    "selected": false,
  },
  Object {
    "number": 2016,
    "selected": false,
  },
  Object {
    "number": 2015,
    "selected": true,
  },
  Object {
    "number": 2014,
    "selected": false,
  },
  Object {
    "number": 2013,
    "selected": false,
  },
  Object {
    "number": 2012,
    "selected": false,
  },
  Object {
    "number": 2011,
    "selected": false,
  },
  Object {
    "number": 2010,
    "selected": false,
  },
  Object {
    "number": 2009,
    "selected": false,
  },
  Object {
    "number": 2008,
    "selected": false,
  },
  Object {
    "number": 2007,
    "selected": false,
  },
  Object {
    "number": 2006,
    "selected": false,
  },
  Object {
    "number": 2005,
    "selected": false,
  },
  Object {
    "number": 2004,
    "selected": false,
  },
  Object {
    "number": 2003,
    "selected": false,
  },
  Object {
    "number": 2002,
    "selected": false,
  },
  Object {
    "number": 2001,
    "selected": false,
  },
  Object {
    "number": 2000,
    "selected": false,
  },
]
`;

exports[`DatePickerComponent generating years should max at tracked year when tracked is later 1`] = `
Array [
  Object {
    "number": 2040,
    "selected": true,
  },
  Object {
    "number": 2039,
    "selected": false,
  },
  Object {
    "number": 2038,
    "selected": false,
  },
  Object {
    "number": 2037,
    "selected": false,
  },
  Object {
    "number": 2036,
    "selected": false,
  },
  Object {
    "number": 2035,
    "selected": false,
  },
  Object {
    "number": 2034,
    "selected": false,
  },
  Object {
    "number": 2033,
    "selected": false,
  },
  Object {
    "number": 2032,
    "selected": false,
  },
  Object {
    "number": 2031,
    "selected": false,
  },
  Object {
    "number": 2030,
    "selected": false,
  },
  Object {
    "number": 2029,
    "selected": false,
  },
  Object {
    "number": 2028,
    "selected": false,
  },
  Object {
    "number": 2027,
    "selected": false,
  },
  Object {
    "number": 2026,
    "selected": false,
  },
  Object {
    "number": 2025,
    "selected": false,
  },
  Object {
    "number": 2024,
    "selected": false,
  },
  Object {
    "number": 2023,
    "selected": false,
  },
  Object {
    "number": 2022,
    "selected": false,
  },
  Object {
    "number": 2021,
    "selected": false,
  },
  Object {
    "number": 2020,
    "selected": false,
  },
  Object {
    "number": 2019,
    "selected": false,
  },
  Object {
    "number": 2018,
    "selected": false,
  },
  Object {
    "number": 2017,
    "selected": false,
  },
  Object {
    "number": 2016,
    "selected": false,
  },
  Object {
    "number": 2015,
    "selected": false,
  },
  Object {
    "number": 2014,
    "selected": false,
  },
  Object {
    "number": 2013,
    "selected": false,
  },
  Object {
    "number": 2012,
    "selected": false,
  },
  Object {
    "number": 2011,
    "selected": false,
  },
  Object {
    "number": 2010,
    "selected": false,
  },
  Object {
    "number": 2009,
    "selected": false,
  },
  Object {
    "number": 2008,
    "selected": false,
  },
  Object {
    "number": 2007,
    "selected": false,
  },
  Object {
    "number": 2006,
    "selected": false,
  },
  Object {
    "number": 2005,
    "selected": false,
  },
  Object {
    "number": 2004,
    "selected": false,
  },
  Object {
    "number": 2003,
    "selected": false,
  },
  Object {
    "number": 2002,
    "selected": false,
  },
  Object {
    "number": 2001,
    "selected": false,
  },
  Object {
    "number": 2000,
    "selected": false,
  },
]
`;

exports[`DatePickerComponent generating years should min at 2000 year when tracked is later 1`] = `
Array [
  Object {
    "number": 2022,
    "selected": false,
  },
  Object {
    "number": 2021,
    "selected": false,
  },
  Object {
    "number": 2020,
    "selected": false,
  },
  Object {
    "number": 2019,
    "selected": false,
  },
  Object {
    "number": 2018,
    "selected": false,
  },
  Object {
    "number": 2017,
    "selected": false,
  },
  Object {
    "number": 2016,
    "selected": false,
  },
  Object {
    "number": 2015,
    "selected": false,
  },
  Object {
    "number": 2014,
    "selected": false,
  },
  Object {
    "number": 2013,
    "selected": false,
  },
  Object {
    "number": 2012,
    "selected": false,
  },
  Object {
    "number": 2011,
    "selected": false,
  },
  Object {
    "number": 2010,
    "selected": false,
  },
  Object {
    "number": 2009,
    "selected": false,
  },
  Object {
    "number": 2008,
    "selected": false,
  },
  Object {
    "number": 2007,
    "selected": false,
  },
  Object {
    "number": 2006,
    "selected": false,
  },
  Object {
    "number": 2005,
    "selected": false,
  },
  Object {
    "number": 2004,
    "selected": false,
  },
  Object {
    "number": 2003,
    "selected": false,
  },
  Object {
    "number": 2002,
    "selected": true,
  },
  Object {
    "number": 2001,
    "selected": false,
  },
  Object {
    "number": 2000,
    "selected": false,
  },
]
`;

exports[`DatePickerComponent generating years should min at tracked year when tracked is earlier 1`] = `
Array [
  Object {
    "number": 2022,
    "selected": false,
  },
  Object {
    "number": 2021,
    "selected": false,
  },
  Object {
    "number": 2020,
    "selected": false,
  },
  Object {
    "number": 2019,
    "selected": false,
  },
  Object {
    "number": 2018,
    "selected": false,
  },
  Object {
    "number": 2017,
    "selected": false,
  },
  Object {
    "number": 2016,
    "selected": false,
  },
  Object {
    "number": 2015,
    "selected": false,
  },
  Object {
    "number": 2014,
    "selected": false,
  },
  Object {
    "number": 2013,
    "selected": false,
  },
  Object {
    "number": 2012,
    "selected": false,
  },
  Object {
    "number": 2011,
    "selected": false,
  },
  Object {
    "number": 2010,
    "selected": false,
  },
  Object {
    "number": 2009,
    "selected": false,
  },
  Object {
    "number": 2008,
    "selected": false,
  },
  Object {
    "number": 2007,
    "selected": false,
  },
  Object {
    "number": 2006,
    "selected": false,
  },
  Object {
    "number": 2005,
    "selected": false,
  },
  Object {
    "number": 2004,
    "selected": false,
  },
  Object {
    "number": 2003,
    "selected": false,
  },
  Object {
    "number": 2002,
    "selected": false,
  },
  Object {
    "number": 2001,
    "selected": false,
  },
  Object {
    "number": 2000,
    "selected": false,
  },
  Object {
    "number": 1999,
    "selected": false,
  },
  Object {
    "number": 1998,
    "selected": true,
  },
]
`;

exports[`DatePickerComponent should generate months 1`] = `
Array [
  Object {
    "name": "January",
    "selected": false,
  },
  Object {
    "name": "February",
    "selected": false,
  },
  Object {
    "name": "March",
    "selected": false,
  },
  Object {
    "name": "April",
    "selected": false,
  },
  Object {
    "name": "May",
    "selected": false,
  },
  Object {
    "name": "June",
    "selected": false,
  },
  Object {
    "name": "July",
    "selected": true,
  },
  Object {
    "name": "August",
    "selected": false,
  },
  Object {
    "name": "September",
    "selected": false,
  },
  Object {
    "name": "October",
    "selected": false,
  },
  Object {
    "name": "November",
    "selected": false,
  },
  Object {
    "name": "December",
    "selected": false,
  },
]
`;

exports[`DatePickerComponent should generate weeks 1`] = `
Array [
  Array [
    Object {
      "date": 2022-06-26T04:00:00.000Z,
      "number": 26,
      "otherMonth": true,
      "selected": false,
    },
    Object {
      "date": 2022-06-27T04:00:00.000Z,
      "number": 27,
      "otherMonth": true,
      "selected": false,
    },
    Object {
      "date": 2022-06-28T04:00:00.000Z,
      "number": 28,
      "otherMonth": true,
      "selected": false,
    },
    Object {
      "date": 2022-06-29T04:00:00.000Z,
      "number": 29,
      "otherMonth": true,
      "selected": false,
    },
    Object {
      "date": 2022-06-30T04:00:00.000Z,
      "number": 30,
      "otherMonth": true,
      "selected": false,
    },
    Object {
      "date": 2022-07-01T04:00:00.000Z,
      "number": 1,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-02T04:00:00.000Z,
      "number": 2,
      "otherMonth": false,
      "selected": false,
    },
  ],
  Array [
    Object {
      "date": 2022-07-03T04:00:00.000Z,
      "number": 3,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-04T04:00:00.000Z,
      "number": 4,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-05T04:00:00.000Z,
      "number": 5,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-06T04:00:00.000Z,
      "number": 6,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-07T04:00:00.000Z,
      "number": 7,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-08T04:00:00.000Z,
      "number": 8,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-09T04:00:00.000Z,
      "number": 9,
      "otherMonth": false,
      "selected": false,
    },
  ],
  Array [
    Object {
      "date": 2022-07-10T04:00:00.000Z,
      "number": 10,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-11T04:00:00.000Z,
      "number": 11,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-12T04:00:00.000Z,
      "number": 12,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-13T04:00:00.000Z,
      "number": 13,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-14T04:00:00.000Z,
      "number": 14,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-15T04:00:00.000Z,
      "number": 15,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-16T04:00:00.000Z,
      "number": 16,
      "otherMonth": false,
      "selected": false,
    },
  ],
  Array [
    Object {
      "date": 2022-07-17T04:00:00.000Z,
      "number": 17,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-18T04:00:00.000Z,
      "number": 18,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-19T04:00:00.000Z,
      "number": 19,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-20T04:00:00.000Z,
      "number": 20,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-21T04:00:00.000Z,
      "number": 21,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-22T04:00:00.000Z,
      "number": 22,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-23T04:00:00.000Z,
      "number": 23,
      "otherMonth": false,
      "selected": false,
    },
  ],
  Array [
    Object {
      "date": 2022-07-24T04:00:00.000Z,
      "number": 24,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-25T04:00:00.000Z,
      "number": 25,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-26T04:00:00.000Z,
      "number": 26,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-27T04:00:00.000Z,
      "number": 27,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-28T04:00:00.000Z,
      "number": 28,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-29T04:00:00.000Z,
      "number": 29,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-07-30T04:00:00.000Z,
      "number": 30,
      "otherMonth": false,
      "selected": false,
    },
  ],
  Array [
    Object {
      "date": 2022-07-31T04:00:00.000Z,
      "number": 31,
      "otherMonth": false,
      "selected": false,
    },
    Object {
      "date": 2022-08-01T04:00:00.000Z,
      "number": 1,
      "otherMonth": true,
      "selected": false,
    },
    Object {
      "date": 2022-08-02T04:00:00.000Z,
      "number": 2,
      "otherMonth": true,
      "selected": false,
    },
    Object {
      "date": 2022-08-03T04:00:00.000Z,
      "number": 3,
      "otherMonth": true,
      "selected": false,
    },
    Object {
      "date": 2022-08-04T04:00:00.000Z,
      "number": 4,
      "otherMonth": true,
      "selected": false,
    },
    Object {
      "date": 2022-08-05T04:00:00.000Z,
      "number": 5,
      "otherMonth": true,
      "selected": false,
    },
    Object {
      "date": 2022-08-06T04:00:00.000Z,
      "number": 6,
      "otherMonth": true,
      "selected": false,
    },
  ],
]
`;
